# Ultralytics Framework for YOLOv9 Training

In [1]:
import torch

# Number of GPUs available
num_gpus = torch.cuda.device_count()

print(f"Number of available GPUs: {num_gpus}")


Number of available GPUs: 4


# Yolov09m Model Training on:
- Carla_Labelling-v5 dataset
- Train: 70% = 4485 Images
- Validation: 20% = 1274 Images
- Test: 10% = 641 Images

- Hyperparameters are on automatic settings using Ultralytics package

In [2]:
from ultralytics import YOLO

# Load YOLOv10n model from scratch
model = YOLO("yolov9m.pt")

100%|██████████| 39.1M/39.1M [00:00<00:00, 200MB/s]


# Validate the pretrained weights by running inference on an image.

In [3]:
# Run batched inference on a list of images
results = model(["/home/sur06423/project/1.jpg"])  # return a list of Results objects

# Process results list
for result in results:
    boxes = result.boxes  # Boxes object for bounding box outputs
    masks = result.masks  # Masks object for segmentation masks outputs
    keypoints = result.keypoints  # Keypoints object for pose outputs
    probs = result.probs  # Probs object for classification outputs
    obb = result.obb  # Oriented boxes object for OBB outputs
    result.show()  # display to screen
    result.save(filename="result.jpg")  # save to disk


0: 448x640 4 horses, 81.1ms
Speed: 8.7ms preprocess, 81.1ms inference, 461.2ms postprocess per image at shape (1, 3, 448, 640)


Error: no "view" rule for type "image/png" passed its test case
       (for more information, add "--debug=1" on the command line)


# Train the model using the following guidelines:
- Link for various arguments: https://docs.ultralytics.com/modes/train/#train-settings 

In [4]:
import os

# Set library paths
os.environ['LD_LIBRARY_PATH'] = '/usr/lib/xorg-nvidia-525.116.04/lib/x86_64-linux-gnu:' + os.environ.get('LD_LIBRARY_PATH', '')
os.environ['LD_LIBRARY_PATH'] = '/usr/lib/xorg/lib/x86_64-linux-gnu:' + os.environ['LD_LIBRARY_PATH']
os.environ['LD_LIBRARY_PATH'] = '/usr/lib/xorg-nvidia-535.113.01/lib/x86_64-linux-gnu:' + os.environ['LD_LIBRARY_PATH']

# Verify the update
print(os.environ['LD_LIBRARY_PATH'])


/usr/lib/xorg-nvidia-535.113.01/lib/x86_64-linux-gnu:/usr/lib/xorg/lib/x86_64-linux-gnu:/usr/lib/xorg-nvidia-525.116.04/lib/x86_64-linux-gnu:/home/sur06423/miniconda3/envs/yolov_env/lib/python3.9/site-packages/cv2/../../lib64:


In [5]:
# Train the model: https://docs.ultralytics.com/modes/train/#train-settings
model.train(data="/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml", # 	Path to the dataset configuration file (e.g., coco8.yaml). This file contains dataset-specific parameters, including paths to training and validation data, class names, and number of classes.
            epochs=100, # Total number of training epochs. Each epoch represents a full pass over the entire dataset. Adjusting this value can affect training duration and model performance.
            patience=10, # Number of epochs to wait without improvement in validation metrics before early stopping the training. Helps prevent overfitting by stopping training when performance plateaus.
            batch=128, # Batch size, with three modes: set as an integer (e.g., batch=16), auto mode for 60% GPU memory utilization (batch=-1), or auto mode with specified utilization fraction (batch=0.70).
            imgsz=640, # Target image size for training. All images are resized to this dimension before being fed into the model. Affects model accuracy and computational complexity.
            save=True, # Enables saving of training checkpoints and final model weights. Useful for resuming training or model deployment.
            save_period = 1, # Frequency of saving model checkpoints, specified in epochs. A value of -1 disables this feature. Useful for saving interim models during long training sessions.
            device=[0,1,2,3], # Specifies the computational device(s) for training: a single GPU (device=0), multiple GPUs (device=0,1), CPU (device=cpu), or MPS for Apple silicon (device=mps).
            workers=6, # Number of worker threads for data loading (per RANK if Multi-GPU training). Influences the speed of data preprocessing and feeding into the model, especially useful in multi-GPU setups.
            project="/home/sur06423/project/Version_5_exp/YOLO_v9/v9_m", # Name of the project directory where training outputs are saved. Allows for organized storage of different experiments.
            pretrained = True, # Determines whether to start training from a pretrained model. Can be a boolean value or a string path to a specific model from which to load weights. Enhances training efficiency and model performance.
            optimizer = "auto", # Choice of optimizer for training. Options include SGD, Adam, AdamW, NAdam, RAdam, RMSProp etc., or auto for automatic selection based on model configuration. Affects convergence speed and stability.
            verbose = True, # Enables verbose output during training, providing detailed logs and progress updates. Useful for debugging and closely monitoring the training process.
            seed = 0, # Sets the random seed for training, ensuring reproducibility of results across runs with the same configurations.
            amp = True, # Enables Automatic Mixed Precision (AMP) training, reducing memory usage and possibly speeding up training with minimal impact on accuracy.
            profile =True, # Enables profiling of ONNX and TensorRT speeds during training, useful for optimizing model deployment.
            val=True, # Enables validation during training, allowing for periodic evaluation of model performance on a separate dataset.
            plots = True, #Generates and saves plots of training and validation metrics, as well as prediction examples, providing visual insights into model performance and learning progression.                
)

New https://pypi.org/project/ultralytics/8.3.13 available 😃 Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.2.93 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:1 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:2 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:3 (NVIDIA RTX A6000, 48677MiB)


engine/trainer: task=detect, mode=train, model=yolov9m.pt, data=/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml, epochs=100, time=None, patience=10, batch=128, imgsz=640, save=True, save_period=1, cache=False, device=[0, 1, 2, 3], workers=6, project=/home/sur06423/project/Version_5_exp/YOLO_v9/v9_m, name=train, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=True, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, sho

100%|██████████| 6.25M/6.25M [00:00<00:00, 170MB/s]


AMP: checks passed ✅


train: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/train/labels.cache... 4485 images, 430 backgrounds, 0 corrupt: 100%|██████████| 4485/4485 [00:00<?, ?it/s]
val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/valid/labels.cache... 1274 images, 124 backgrounds, 0 corrupt: 100%|██████████| 1274/1274 [00:00<?, ?it/s]


Plotting labels to /home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000714, momentum=0.9) with parameter groups 149 weight(decay=0.0), 156 weight(decay=0.001), 155 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 24 dataloader workers
Logging results to /home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train
Starting training for 100 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      1/100      16.6G      1.029      2.364      1.072          5        640: 100%|██████████| 36/36 [00:19<00:00,  1.84it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.52it/s]


                   all       1274       7377      0.738      0.441      0.481      0.278

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      2/100      16.7G     0.9117     0.7436     0.9858         17        640: 100%|██████████| 36/36 [00:17<00:00,  2.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.709       0.42      0.453      0.273

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      3/100      16.5G     0.9633     0.7827      1.018         37        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.61it/s]


                   all       1274       7377      0.446      0.357      0.267      0.141

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      4/100      16.5G      0.955     0.7687      1.022          8        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.62it/s]


                   all       1274       7377      0.404      0.425      0.295      0.149

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      5/100      16.5G     0.9475     0.7148      1.018          6        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.709      0.464      0.494      0.282

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      6/100      16.5G     0.9149     0.6489     0.9907         26        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.61it/s]


                   all       1274       7377      0.664      0.483      0.503       0.31

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      7/100      16.7G     0.8881     0.6334     0.9825          3        640: 100%|██████████| 36/36 [00:17<00:00,  2.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.62it/s]


                   all       1274       7377      0.773      0.494      0.556       0.34

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      8/100      16.4G     0.8751     0.6006     0.9633         24        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.787      0.488      0.556      0.341

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      9/100      16.7G     0.8369     0.5751     0.9669          1        640: 100%|██████████| 36/36 [00:17<00:00,  2.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.60it/s]


                   all       1274       7377       0.79      0.528      0.588      0.352

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     10/100      16.5G     0.8468     0.5574     0.9529          8        640: 100%|██████████| 36/36 [00:17<00:00,  2.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.792      0.522      0.592      0.367

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     11/100      16.5G     0.8546     0.5566     0.9513         14        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.59it/s]


                   all       1274       7377      0.781      0.533      0.594       0.36

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     12/100      16.4G     0.8409     0.5643     0.9365          1        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.62it/s]


                   all       1274       7377      0.839      0.542      0.611      0.369

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     13/100      16.7G     0.8233     0.5271     0.9452         11        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.62it/s]


                   all       1274       7377      0.803      0.556      0.611      0.374

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     14/100      16.5G     0.8119      0.538     0.9344         14        640: 100%|██████████| 36/36 [00:17<00:00,  2.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.62it/s]


                   all       1274       7377      0.841      0.546      0.629      0.392

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     15/100      16.5G     0.7984     0.5092     0.9329         24        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.813      0.548       0.61      0.384

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     16/100      16.7G      0.797     0.5198     0.9386          4        640: 100%|██████████| 36/36 [00:17<00:00,  2.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.812      0.578      0.639      0.401

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     17/100      16.7G     0.8071     0.5244     0.9503         15        640: 100%|██████████| 36/36 [00:17<00:00,  2.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.821      0.572      0.639      0.396

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     18/100      16.5G      0.782     0.5131     0.9467          6        640: 100%|██████████| 36/36 [00:16<00:00,  2.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.839      0.569       0.64      0.394

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     19/100      16.7G     0.7798     0.4806     0.9288         12        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.62it/s]


                   all       1274       7377      0.843      0.559      0.641      0.402

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     20/100      16.7G     0.7824     0.4881     0.9376         23        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.812      0.585      0.655      0.419

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     21/100      16.7G     0.7584     0.4712     0.9273          4        640: 100%|██████████| 36/36 [00:17<00:00,  2.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.839      0.588      0.658       0.41

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     22/100      16.7G     0.7881     0.4912     0.9295          3        640: 100%|██████████| 36/36 [00:17<00:00,  2.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.833      0.587      0.655      0.417

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     23/100      16.7G     0.7461      0.459     0.9186         27        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.836      0.587      0.663      0.428

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     24/100      16.4G     0.7402     0.4548     0.9183          3        640: 100%|██████████| 36/36 [00:17<00:00,  2.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.833      0.611      0.682      0.438

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     25/100      16.7G     0.7208     0.4559     0.9114         20        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.857      0.593      0.673      0.432

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     26/100      16.7G     0.7232     0.4385     0.9269         18        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.834      0.607      0.672      0.432

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     27/100      16.7G     0.7536     0.4495     0.9143         11        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.844      0.601      0.677      0.426

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     28/100      16.5G     0.7188     0.4387     0.9085         20        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.853      0.608      0.681       0.44

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     29/100      16.7G     0.7025     0.4222     0.8997          7        640: 100%|██████████| 36/36 [00:17<00:00,  2.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.61it/s]


                   all       1274       7377       0.85      0.601      0.682      0.442

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     30/100      16.5G     0.7251     0.4355     0.9171         11        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.60it/s]


                   all       1274       7377      0.841      0.611      0.682      0.435

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     31/100      16.7G     0.7015       0.43     0.9059         19        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.55it/s]


                   all       1274       7377      0.842      0.612      0.679      0.432

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     32/100      16.7G     0.7079     0.4228     0.9017         10        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.62it/s]


                   all       1274       7377      0.855      0.613      0.687      0.443

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     33/100      16.4G     0.6881     0.4093      0.901         13        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.843      0.619      0.693      0.451

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     34/100      16.7G     0.7056     0.4234     0.9106         11        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.62it/s]


                   all       1274       7377      0.857      0.607      0.689       0.45

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     35/100      16.7G     0.7082      0.415     0.9149         21        640: 100%|██████████| 36/36 [00:17<00:00,  2.04it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:08<00:00,  2.36it/s]


                   all       1274       7377      0.845      0.626      0.696      0.454

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     36/100      16.7G     0.6868     0.4098     0.8993          9        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.62it/s]


                   all       1274       7377      0.846      0.611      0.686      0.448

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     37/100      16.5G     0.6967     0.4115     0.9064         10        640: 100%|██████████| 36/36 [00:17<00:00,  2.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:08<00:00,  2.38it/s]


                   all       1274       7377      0.876      0.625      0.707      0.465

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     38/100      16.5G     0.6614     0.4996     0.8834          0        640: 100%|██████████| 36/36 [00:17<00:00,  2.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377       0.84      0.632        0.7      0.455

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     39/100      16.7G     0.6696     0.3933      0.897         21        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.839      0.629      0.697      0.451

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     40/100      16.5G     0.6763     0.4009     0.8956         15        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.62it/s]


                   all       1274       7377      0.849      0.642      0.711      0.466

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     41/100      16.5G      0.674     0.4045     0.9049         10        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.58it/s]


                   all       1274       7377      0.861      0.626      0.707      0.462

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     42/100      16.7G     0.6636     0.3925     0.8952         24        640: 100%|██████████| 36/36 [00:17<00:00,  2.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377       0.86      0.638       0.71      0.465

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     43/100      16.7G     0.6903     0.4305     0.9216          4        640: 100%|██████████| 36/36 [00:17<00:00,  2.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.845      0.638      0.709       0.47

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     44/100      16.5G     0.6815      0.389     0.8858         26        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.864      0.641      0.714      0.469

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     45/100      16.7G     0.6495     0.3786     0.8954          8        640: 100%|██████████| 36/36 [00:17<00:00,  2.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.852      0.644      0.718      0.467

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     46/100      16.7G     0.6371     0.3702     0.8914          8        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.857       0.64      0.719      0.474

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     47/100      16.5G     0.6427     0.3754     0.8886         15        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.842      0.643      0.712      0.472

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     48/100      16.5G      0.654     0.3716      0.887         16        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.869      0.641      0.711      0.465

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     49/100      16.5G     0.6805     0.4978     0.8718          7        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.62it/s]


                   all       1274       7377      0.858      0.645      0.719      0.472

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     50/100      16.5G     0.6609     0.3794     0.8833         17        640: 100%|██████████| 36/36 [00:17<00:00,  2.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.864      0.645       0.72      0.476

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     51/100      16.7G     0.6499     0.3731     0.8986         10        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.859      0.649      0.724      0.477

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     52/100      16.7G     0.6305     0.3617     0.8888          5        640: 100%|██████████| 36/36 [00:17<00:00,  2.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377       0.89      0.628      0.722      0.474

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     53/100      16.5G     0.6377     0.3677     0.8924          8        640: 100%|██████████| 36/36 [00:17<00:00,  2.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:08<00:00,  2.45it/s]


                   all       1274       7377      0.856      0.645      0.722      0.477

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     54/100      16.5G     0.6111     0.3545     0.8837         14        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.869      0.633      0.717      0.477

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     55/100      16.7G     0.6327     0.3755     0.8835         21        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.877      0.645      0.733      0.488

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     56/100      16.7G     0.6085     0.3517     0.8862          5        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.868      0.646      0.725       0.48

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     57/100      16.7G     0.6245     0.3574     0.8754         13        640: 100%|██████████| 36/36 [00:17<00:00,  2.07it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:08<00:00,  2.37it/s]


                   all       1274       7377      0.827      0.665      0.725      0.479

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     58/100      16.7G     0.6184     0.3517     0.8877         21        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.851      0.648      0.729      0.483

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     59/100      16.5G     0.6192     0.3456     0.8777         12        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.852       0.66      0.733      0.491

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     60/100      16.5G     0.6177     0.3516     0.8796         18        640: 100%|██████████| 36/36 [00:17<00:00,  2.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.66it/s]


                   all       1274       7377      0.865      0.651      0.733      0.494

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     61/100      16.5G     0.6197     0.3522     0.8778         39        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.66it/s]


                   all       1274       7377      0.873      0.651      0.739      0.492

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     62/100      16.5G     0.5986     0.3417     0.8789         33        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.882      0.638      0.736      0.493

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     63/100      16.5G     0.6203     0.3509     0.8762         13        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.866      0.653      0.735      0.491

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     64/100      16.6G     0.6256     0.3572     0.8717          6        640: 100%|██████████| 36/36 [00:17<00:00,  2.07it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.874      0.644      0.729       0.49

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     65/100      16.7G     0.6387     0.3481     0.8795          3        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.897      0.636      0.739      0.499

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     66/100      16.5G     0.6141     0.3412      0.886         10        640: 100%|██████████| 36/36 [00:17<00:00,  2.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.884      0.654      0.736       0.49

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     67/100      16.7G     0.6169     0.3463     0.8807         19        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.856      0.669      0.742      0.499

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     68/100      16.7G     0.5893     0.3327     0.8654         19        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.883      0.651      0.739      0.494

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     69/100      16.7G     0.6141      0.339     0.8744         16        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.865      0.664      0.746      0.502

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     70/100      16.5G     0.5998     0.3326     0.8734         24        640: 100%|██████████| 36/36 [00:17<00:00,  2.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.879      0.658      0.749      0.503

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     71/100      16.7G     0.5886     0.3194     0.8743         17        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.858      0.659      0.739      0.502

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     72/100      16.4G     0.5867     0.3223      0.872         12        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.875      0.652      0.745      0.505

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     73/100      16.5G     0.5941     0.3302      0.879         13        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.867      0.662      0.741      0.499

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     74/100      16.7G      0.574     0.3189     0.8802         11        640: 100%|██████████| 36/36 [00:17<00:00,  2.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.868       0.65      0.739      0.502

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     75/100      16.7G     0.5877     0.3184     0.8699          6        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.62it/s]


                   all       1274       7377      0.858      0.664      0.749      0.505

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     76/100      16.7G     0.5901     0.3193     0.8645          8        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.869      0.657      0.745      0.502

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     77/100      16.6G     0.5871     0.3184     0.8685         22        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.66it/s]


                   all       1274       7377      0.882      0.654      0.741      0.499

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     78/100      16.7G     0.5843     0.3114      0.881          4        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.881       0.66      0.751      0.506

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     79/100      16.5G     0.5724     0.3118     0.8695         13        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.881      0.664      0.755      0.513

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     80/100      16.7G     0.5643     0.3101     0.8649         36        640: 100%|██████████| 36/36 [00:17<00:00,  2.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377       0.88      0.658      0.744      0.505

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     81/100      16.7G     0.5637     0.3082     0.8695          7        640: 100%|██████████| 36/36 [00:17<00:00,  2.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.873      0.664      0.748       0.51

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     82/100      16.7G     0.5753     0.3177     0.8694          9        640: 100%|██████████| 36/36 [00:17<00:00,  2.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.879       0.66      0.746      0.506

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     83/100      16.7G     0.5701     0.3058     0.8678          8        640: 100%|██████████| 36/36 [00:17<00:00,  2.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.63it/s]


                   all       1274       7377      0.886      0.657      0.748      0.508

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     84/100      16.5G      0.574     0.3106     0.8676          8        640: 100%|██████████| 36/36 [00:17<00:00,  2.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.876      0.666      0.753       0.51

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     85/100      16.4G     0.5659     0.3029      0.873         11        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.882      0.667      0.753      0.515

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     86/100      16.5G     0.6301     0.3754     0.9646          3        640: 100%|██████████| 36/36 [00:17<00:00,  2.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.867      0.667       0.75       0.51

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     87/100      16.5G     0.5658     0.2986     0.8649         14        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.877      0.661      0.754      0.512

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     88/100      16.7G     0.5615     0.3004      0.863         16        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.882      0.668      0.759      0.518

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     89/100      16.5G     0.5325     0.2825     0.8688         14        640: 100%|██████████| 36/36 [00:17<00:00,  2.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.867      0.674      0.756      0.518

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     90/100      16.7G     0.5356     0.2953     0.8738          3        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.886      0.661       0.76      0.518
Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     91/100      16.7G     0.6064     0.2973     0.8523         13        640: 100%|██████████| 36/36 [00:18<00:00,  2.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:08<00:00,  2.49it/s]


                   all       1274       7377      0.883      0.661      0.752      0.514

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     92/100      16.7G        0.6     0.2965     0.8516         13        640: 100%|██████████| 36/36 [00:17<00:00,  2.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.888       0.66      0.756      0.515

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     93/100      16.5G     0.5964     0.2956     0.8472          7        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.882      0.664      0.752      0.514

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     94/100      16.7G     0.6065     0.2989     0.8478          6        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.62it/s]


                   all       1274       7377      0.908      0.651      0.755      0.516

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     95/100      16.7G     0.6045     0.2971     0.8449         31        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:08<00:00,  2.49it/s]


                   all       1274       7377      0.899      0.651      0.755      0.515

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     96/100      16.5G     0.5954      0.304     0.8497         13        640: 100%|██████████| 36/36 [00:17<00:00,  2.04it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:08<00:00,  2.39it/s]


                   all       1274       7377      0.861      0.677      0.757      0.519

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     97/100      16.5G     0.5951     0.2903     0.8493         13        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.859      0.676      0.759      0.521

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     98/100      16.5G     0.5931     0.2908     0.8382          3        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.64it/s]


                   all       1274       7377      0.853      0.685      0.763      0.521

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     99/100      16.5G     0.5816     0.2836     0.8425         40        640: 100%|██████████| 36/36 [00:17<00:00,  2.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.855      0.681      0.762      0.521

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


    100/100      16.6G     0.5797     0.2825     0.8511          6        640: 100%|██████████| 36/36 [00:17<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.65it/s]


                   all       1274       7377      0.859      0.681      0.761       0.52

100 epochs completed in 0.933 hours.
Optimizer stripped from /home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train/weights/last.pt, 40.8MB
Optimizer stripped from /home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train/weights/best.pt, 40.8MB

Validating /home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train/weights/best.pt...
Ultralytics YOLOv8.2.93 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:1 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:2 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:3 (NVIDIA RTX A6000, 48677MiB)
YOLOv9m summary (fused): 374 layers, 20,015,884 parameters, 0 gradients, 76.5 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:09<00:00,  2.07it/s]


                   all       1274       7377      0.853      0.685      0.764      0.521
            Pedestrian        333        600      0.758      0.733      0.772      0.474
         Traffic_Signs        309        434      0.857      0.629      0.725      0.441
               Vehicle       1034       2820      0.923      0.905      0.946       0.78
         traffic_light        660       3523      0.876      0.472      0.612       0.39
Speed: 0.1ms preprocess, 3.2ms inference, 0.0ms loss, 0.7ms postprocess per image
Results saved to /home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train


# Results on Validation set:
100 epochs completed in 0.933 hours.
Optimizer stripped from /home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train/weights/last.pt, 40.8MB
Optimizer stripped from /home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train/weights/best.pt, 40.8MB

Validating /home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train/weights/best.pt...
Ultralytics YOLOv8.2.93 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:1 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:2 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:3 (NVIDIA RTX A6000, 48677MiB)
YOLOv9m summary (fused): 374 layers, 20,015,884 parameters, 0 gradients, 76.5 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:09<00:00,  2.07it/s]
                   all       1274       7377      0.853      0.685      0.764      0.521
            Pedestrian        333        600      0.758      0.733      0.772      0.474
         Traffic_Signs        309        434      0.857      0.629      0.725      0.441
               Vehicle       1034       2820      0.923      0.905      0.946       0.78
         traffic_light        660       3523      0.876      0.472      0.612       0.39
Speed: 0.1ms preprocess, 3.2ms inference, 0.0ms loss, 0.7ms postprocess per image
Results saved to /home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train

# Test the results on Test dataset
- The above results was on the validation set of the Carla_labeling-v5 dataset (See the Image counts)
- Now we need to calculate the metrics on the test set of the Carla_labeling-v5 dataset containing only 641 images.
- Below is the script to do so:

In [6]:
from ultralytics import YOLO

# Load YOLOv10s model from scratch
# model = YOLO("yolov10s.pt")
# Load YOLOv10s Best weights on carla dataset
model = YOLO("/home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train/weights/last.pt")

# Train the model: https://docs.ultralytics.com/modes/train/#train-settings
validation_results = model.val(data="/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml", 
            batch=128, 
            imgsz=640,
            save_json = True, # If True, saves the results to a JSON file for further analysis or integration with other tools.
            conf=0.25, # Sets the minimum confidence threshold for detections. Detections with confidence below this threshold are discarded.
            iou=0.6, # Sets the Intersection Over Union (IoU) threshold for Non-Maximum Suppression (NMS). Helps in reducing duplicate detections.
            device=0, 
            split="test",              
)

Ultralytics YOLOv8.2.93 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
YOLOv9m summary (fused): 374 layers, 20,015,884 parameters, 0 gradients, 76.5 GFLOPs


val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/test/labels.cache... 641 images, 53 backgrounds, 0 corrupt: 100%|██████████| 641/641 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:12<00:00,  2.00s/it]


                   all        641       3701      0.896      0.649      0.795      0.586
            Pedestrian        182        332      0.796      0.657      0.775      0.526
         Traffic_Signs        168        230      0.947      0.617      0.791      0.538
               Vehicle        528       1377      0.947      0.882      0.935       0.81
         traffic_light        338       1762      0.895       0.44      0.679      0.471
Speed: 0.4ms preprocess, 5.9ms inference, 0.0ms loss, 1.5ms postprocess per image
Saving runs/detect/val/predictions.json...
Results saved to runs/detect/val


# Results on Last checpoint (100th epoch):
Ultralytics YOLOv8.2.93 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
YOLOv9m summary (fused): 374 layers, 20,015,884 parameters, 0 gradients, 76.5 GFLOPs
val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/test/labels.cache... 641 images, 53 backgrounds, 0 corrupt: 100%|██████████| 641/641 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:12<00:00,  2.00s/it]
                   all        641       3701      0.896      0.649      0.795      0.586
            Pedestrian        182        332      0.796      0.657      0.775      0.526
         Traffic_Signs        168        230      0.947      0.617      0.791      0.538
               Vehicle        528       1377      0.947      0.882      0.935       0.81
         traffic_light        338       1762      0.895       0.44      0.679      0.471
Speed: 0.4ms preprocess, 5.9ms inference, 0.0ms loss, 1.5ms postprocess per image
Saving runs/detect/val/predictions.json...
Results saved to runs/detect/val

# Results on Test Set using Best Checkpoint:

In [7]:
from ultralytics import YOLO

# Load YOLOv10s model from scratch
# model = YOLO("yolov10s.pt")
# Load the model
model = YOLO("/home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train/weights/best.pt")

# Train the model: https://docs.ultralytics.com/modes/train/#train-settings
validation_results = model.val(data="/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml", 
            batch=128, 
            imgsz=640,
            save_json = True, # If True, saves the results to a JSON file for further analysis or integration with other tools.
            conf=0.25, # Sets the minimum confidence threshold for detections. Detections with confidence below this threshold are discarded.
            iou=0.6, # Sets the Intersection Over Union (IoU) threshold for Non-Maximum Suppression (NMS). Helps in reducing duplicate detections.
            device=0, 
            split="test",              
)

Ultralytics YOLOv8.2.93 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
YOLOv9m summary (fused): 374 layers, 20,015,884 parameters, 0 gradients, 76.5 GFLOPs


val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/test/labels.cache... 641 images, 53 backgrounds, 0 corrupt: 100%|██████████| 641/641 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:11<00:00,  1.95s/it]


                   all        641       3701      0.897      0.648      0.795      0.586
            Pedestrian        182        332      0.815      0.663      0.781      0.532
         Traffic_Signs        168        230      0.933      0.609      0.785      0.534
               Vehicle        528       1377      0.946      0.885      0.937       0.81
         traffic_light        338       1762      0.894      0.436      0.678      0.468
Speed: 0.3ms preprocess, 5.2ms inference, 0.0ms loss, 1.6ms postprocess per image
Saving runs/detect/val2/predictions.json...
Results saved to runs/detect/val2


# Taking the Optimal value from the F1-Confidence curve for Optimal Confidence
- Optimal Confidence YOLOv9m - 0.156
- IoU threshold : 0.5
- Results save d at: runs/detect/val3

In [1]:
from ultralytics import YOLO

# Load YOLOv10s model from scratch
# model = YOLO("yolov10s.pt")
# Load YOLOv10s Best weights on carla dataset
model = YOLO("/home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train/weights/last.pt")

# Train the model: https://docs.ultralytics.com/modes/train/#train-settings
validation_results = model.val(data="/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml", 
            batch=128, 
            imgsz=640,
            save_json = True, # If True, saves the results to a JSON file for further analysis or integration with other tools.
            conf=0.156, # Sets the minimum confidence threshold for detections. Detections with confidence below this threshold are discarded.
            iou=0.5, # Sets the Intersection Over Union (IoU) threshold for Non-Maximum Suppression (NMS). Helps in reducing duplicate detections.
            device=0, 
            split="test",              
)

WARNING ⚠️ Python>=3.10 is required, but Python==3.9.19 is currently installed 
Ultralytics 8.3.13 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
YOLOv9m summary (fused): 374 layers, 20,015,884 parameters, 0 gradients, 76.5 GFLOPs


val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/test/labels.cache... 641 images, 53 backgrounds, 0 corrupt: 100%|██████████| 641/641 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:09<00:00,  1.55s/it]


                   all        641       3701      0.859       0.68      0.802      0.582
            Pedestrian        182        332      0.753       0.69      0.781      0.521
         Traffic_Signs        168        230       0.89       0.67      0.805      0.532
               Vehicle        528       1377      0.933      0.893       0.94       0.81
         traffic_light        338       1762      0.861      0.466      0.682      0.464
Speed: 2.0ms preprocess, 5.6ms inference, 0.0ms loss, 1.7ms postprocess per image
Saving runs/detect/val3/predictions.json...
Results saved to runs/detect/val3


# Best + Taking the Optimal value from the F1-Confidence curve for Optimal Confidence
- Optimal Confidence YOLOv9m - 0.156
- IoU threshold : 0.5
- Results save d at: runs/detect/val4

In [1]:
from ultralytics import YOLO

# Load YOLOv10s model from scratch
# model = YOLO("yolov10s.pt")
# Load YOLOv10s Best weights on carla dataset
model = YOLO("/home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train/weights/best.pt")

# Train the model: https://docs.ultralytics.com/modes/train/#train-settings
validation_results = model.val(data="/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml", 
            batch=128, 
            imgsz=640,
            save_json = True, # If True, saves the results to a JSON file for further analysis or integration with other tools.
            conf=0.156, # Sets the minimum confidence threshold for detections. Detections with confidence below this threshold are discarded.
            iou=0.5, # Sets the Intersection Over Union (IoU) threshold for Non-Maximum Suppression (NMS). Helps in reducing duplicate detections.
            device=0, 
            split="test",              
)

WARNING ⚠️ Python>=3.10 is required, but Python==3.9.19 is currently installed 
Ultralytics 8.3.13 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
YOLOv9m summary (fused): 374 layers, 20,015,884 parameters, 0 gradients, 76.5 GFLOPs


val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/test/labels.cache... 641 images, 53 backgrounds, 0 corrupt: 100%|██████████| 641/641 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:08<00:00,  1.41s/it]


                   all        641       3701      0.859      0.674        0.8      0.581
            Pedestrian        182        332      0.763      0.687      0.783      0.523
         Traffic_Signs        168        230      0.877      0.652      0.794      0.527
               Vehicle        528       1377      0.932      0.893       0.94       0.81
         traffic_light        338       1762      0.865      0.465      0.682      0.464
Speed: 1.8ms preprocess, 5.7ms inference, 0.0ms loss, 1.8ms postprocess per image
Saving runs/detect/val4/predictions.json...
Results saved to runs/detect/val4


# Best Epoch 98 +  Optimal Confidence + IoU = 0.6
- Optimal Confidence YOLOv9m - 0.156
- IoU threshold : 0.6
- Results save d at: runs/detect/val5

In [1]:
from ultralytics import YOLO

# Load YOLOv10s model from scratch
# model = YOLO("yolov10s.pt")
# Load YOLOv10s Best weights on carla dataset
model = YOLO("/home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train/weights/best.pt")

# Train the model: https://docs.ultralytics.com/modes/train/#train-settings
validation_results = model.val(data="/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml", 
            batch=128, 
            imgsz=640,
            save_json = True, # If True, saves the results to a JSON file for further analysis or integration with other tools.
            conf=0.156, # Sets the minimum confidence threshold for detections. Detections with confidence below this threshold are discarded.
            iou=0.6, # Sets the Intersection Over Union (IoU) threshold for Non-Maximum Suppression (NMS). Helps in reducing duplicate detections.
            device=0, 
            split="test",              
)

WARNING ⚠️ Python>=3.10 is required, but Python==3.9.19 is currently installed 
Ultralytics 8.3.13 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
YOLOv9m summary (fused): 374 layers, 20,015,884 parameters, 0 gradients, 76.5 GFLOPs


val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/test/labels.cache... 641 images, 53 backgrounds, 0 corrupt: 100%|██████████| 641/641 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:08<00:00,  1.40s/it]


                   all        641       3701      0.853      0.674      0.799       0.58
            Pedestrian        182        332       0.75      0.687       0.78      0.522
         Traffic_Signs        168        230      0.877      0.652      0.794      0.527
               Vehicle        528       1377      0.923      0.894       0.94      0.809
         traffic_light        338       1762      0.862      0.465      0.681      0.463
Speed: 1.9ms preprocess, 5.7ms inference, 0.0ms loss, 1.3ms postprocess per image
Saving runs/detect/val5/predictions.json...
Results saved to runs/detect/val5


# Seperate evaluation on the training set

In [1]:
from ultralytics import YOLO

# Load your model
model = YOLO('/home/sur06423/project/Version_5_exp/YOLO_v9/v9_m/train/weights/best.pt')  # replace with your model path

# Validate on custom dataset
results = model.val(data='/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml',
            batch=128, 
            imgsz=640,
            save_json = True, # If True, saves the results to a JSON file for further analysis or integration with other tools.
            conf=0.25, # Sets the optimal confidence threshold for detections. Detections with confidence below this threshold are discarded.
            iou=0.6, # Sets the Intersection Over Union (IoU) threshold for Non-Maximum Suppression (NMS). Helps in reducing duplicate detections.
            device=0, 
            split="train", 
)  # Ensure your YAML points to your desired dataset

WARNING ⚠️ Python>=3.10 is required, but Python==3.9.19 is currently installed 
Ultralytics 8.3.13 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
YOLOv9m summary (fused): 374 layers, 20,015,884 parameters, 0 gradients, 76.5 GFLOPs


val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/train/labels.cache... 4485 images, 430 backgrounds, 0 corrupt: 100%|██████████| 4485/4485 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 36/36 [00:41<00:00,  1.16s/it]


                   all       4485      26788      0.927      0.694       0.83      0.653
            Pedestrian       1229       2227      0.888      0.754      0.856       0.65
         Traffic_Signs       1093       1459      0.929      0.668      0.817      0.589
               Vehicle       3636      10031      0.964      0.915      0.954      0.854
         traffic_light       2373      13071      0.929      0.438      0.693      0.518
Speed: 0.4ms preprocess, 5.2ms inference, 0.0ms loss, 0.6ms postprocess per image
Saving runs/detect/val6/predictions.json...
Results saved to runs/detect/val6


# This notebook is complete.
- All the cell are executed and saved for later use.
- Results are saved in the corresponding directories.